adr/0008-phasing-out-config-classes.adoc

```diff
@@ -0,0 +1,95 @@
+
+If we phase out the legacy approach, we need the extension ecosystem to be ready for it.
+
+For Quarkiverse extensions, the Quarkus community can be proactive as we can push changes there if needed.
+Things become a bit more blurry for extensions hosted outside of the Quarkiverse for which we will need the maintainers to be doing the work.
+
+=== Compatibility classes
+
+During the migration, for a few key extensions, we decided to keep small and targeted config classes around to maintain the compatibility for all the extensions that consume this configuration (either as a CDI bean at runtime or during build time).
+This is for instance the case of the HTTP configuration coming from Vert.x HTTP.
+
+When we phase out the legacy approach, these classes will have to go: we need to make sure the extension ecosystem has moved to the new interfaces (or to a proper API if we put one in place).
+
+== Proposed approach
+
+Given all the items presented above, the key for this transition to be a success is communication and planning.
+Here is the approach we recommend:
+
+- Step 1 - Communicate first about where we introduced compatibility config classes and that people should consume the new config interface ASAP
+- Step 2 - Communicate early about this upcoming transition
+  - Blog post + post on quarkus-dev
+  - Post in the Quarkiverse community discussions
+  - Prepare a guide on how to do it. Make sure we add useful pointers with example pull requests.
+  - Make sure we provide points of contact in the community for people having difficulties (or who just don't have the time to do it and want someone else to handle it)
+  - Make sure we identify difficulties with potential solutions
+  - Announce specific versions and dates for the tentative phase out so that people can schedule their work
+- Step 3 - In version X.Y (with at least two months between previous communication and release), throw an error in the extension annotation processor when legacy `@ConfigRoot` classes are detected
+  - It will allow us to detect the overall status in Ecosystem CI (at least for public extensions)
+  - Maybe provide a way to forcefully override this error for one version, and drop it in the next version
+  - Keep the support for legacy `@ConfigRoot` classes in quarkus-core though as we want the extensions to still work until they are updated
+  - Announce it via blog post + post on quarkus-dev
+- Step 4 - In version X.(Y + 2), if we are satisfied with the state of the Ecosystem, drop the legacy `@ConfigRoot` classes support from Quarkus entirely
+
+== Proposed schedule
+
+Given we were able to get everything ready for 3.19 and 3.20 LTS, here is the schedule we would suggest:
+
+- Step 1 - For the 3.19 release on February 26th
+- Step 2 - For the 3.19 release - and we should amplify the message when 3.20 LTS gets out
+- Step 3 - For the 3.22 release on April 30th
+- Step 4 - For the 3.24 release on June 25th
```

Comment:
I would add a paragraph like:

This effort is essential. It:

- reduces the codebase complexity
- avoid ambiguities and bad practices
- improves startup performances

Reply:
Done!